New Table: Windows Firewall Rules

aleksmaus opened the pull request:
This change adds the new table ```windows_firewall_rules``` as requested in the issue:  https://github.com/osquery/osquery/issues/7335

There are couple of deviations from the original requirement:
* No ```version``` column, since it's not available through the API.
* Reordered columns slightly.
* Using ```enabled``` column name instead of ```active```, more consistent with the API naming.
* Using INTEGER for ```enabled``` column instead of TRUE/FALSE TEXT value.
* Renamed ```rule_action``` into ```action``` field,  more consistent with the API naming.
* Renamed ```profile``` into ```profiles``` column that contains decoded comma separated text values like ```Private,Public```.  Could create individual rows for each if needed. Let me know.

Screenshot:
<img width="1703" alt="Screen Shot 2021-12-01 at 9 48 39 AM" src="https://user-images.githubusercontent.com/872351/144257333-fcfa1eb3-9fa3-41e1-a068-7f84643f2939.png">

I was not sure if we can use ATL CComPtr wrappers for example, it's cleaner this way. 
Can rewrite without them if needed. Let me know.

Looking for a feedback if there is anything needs to be changed/added/removed.

Fixes: #7335

directionless commented on osquery/tables/networking/windows/defender_firewall_rules.cpp:
```diff
@@ -0,0 +1,191 @@
+    return hr;
+  }
+
+  std::string profiles;
+  if (profileBitmask == NET_FW_PROFILE2_ALL) {
+    r["profiles"] = "Any";
+  } else {
```
> * Renamed `profile` into `profiles` column that contains decoded comma separated text values like `Private,Public`.  Could create individual rows for each if needed. Let me know.

I think this pattern is hard to work with -- It is quite difficult to filter based on these. I would probably make them into boolean columns. Not sure if `any` should be a column, vs just setting true on the others. (Exposing a bitmask would also work, but isn't very friendly)

aleksmaus replied:
yeah, was not sure about this one. sure, can split into three boolean columns:
```profile_domain```, ```profile_private```, ```profile_public```

according to the enum
```
typedef 
enum NET_FW_PROFILE_TYPE2_
    {
        NET_FW_PROFILE2_DOMAIN	= 0x1,
        NET_FW_PROFILE2_PRIVATE	= 0x2,
        NET_FW_PROFILE2_PUBLIC	= 0x4,
        NET_FW_PROFILE2_ALL	= 0x7fffffff
    } 	NET_FW_PROFILE_TYPE2;
```

https://docs.microsoft.com/en-us/windows/win32/api/icftypes/ne-icftypes-net_fw_profile_type2

aleksmaus replied:
something like this:
<img width="1986" alt="Screen Shot 2021-12-01 at 11 18 47 AM" src="https://user-images.githubusercontent.com/872351/144271758-be96c868-fdf3-4deb-89ca-5eb0addc6758.png">

directionless commented on osquery/tables/networking/windows/defender_firewall_rules.cpp:
```diff
@@ -0,0 +1,191 @@
+  if (FAILED(hr = rule->get_Protocol(&protocol))) {
+    return hr;
+  }
+  switch (protocol) {
```
Are there other protocols? Is the `default` case better as "Other" or NULL? (TBH I don't like either)

aleksmaus replied:
at the moment it's 
```
typedef 
enum NET_FW_IP_PROTOCOL_
    {
        NET_FW_IP_PROTOCOL_TCP	= 6,
        NET_FW_IP_PROTOCOL_UDP	= 17,
        NET_FW_IP_PROTOCOL_ANY	= 256
    } 	NET_FW_IP_PROTOCOL;
```
https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/ff736237(v=vs.120)

directionless commented on specs/windows/defender_firewall_rules.table:
```diff
@@ -0,0 +1,15 @@
+table_name("defender_firewall_rules")
+description("Provides the list of firewall rules.")
```
```suggestion
description("Provides the list of Windows Defender firewall rules.")
```

directionless reviewed (commented):
Thanks fo much for the implementation! I'm really excited to see this get into osquery

amalone-scwx commented:
Are these specific to Defender, or does it apply to Windows network firewall rules in-general?

aleksmaus commented:
> Are these specific to Defender, or does it apply to Windows network firewall rules in-general?

This is windows firewall in general, using API defined in:
https://docs.microsoft.com/en-us/windows/win32/api/netfw/

https://en.wikipedia.org/wiki/Windows_Firewall
"Windows Firewall (officially called Windows Defender Firewall in Windows 10), is a firewall component of Microsoft Windows."

amalone-scwx commented:
Maybe `windows_firewall_rules` is a better name, so as not to be confused with "Defender ATP"?  Also, is there a way to get the ports open in the firewall?  Admins will want to know which TCP and UDP ports are open for each rule.

aleksmaus commented:
> Maybe `windows_firewall_rules` is a better name, so as not to be confused with "Defender ATP"? Also, is there a way to get the ports open in the firewall? Admins will want to know which TCP and UDP ports are open for each rule.

Renamed table ```defender_firewall_rules``` to ```windows_firewall_rules```.
Added columns:
* local_addresses
* remote_addresses
* local_ports
* remote_ports
* icmp_types_codes
<img width="2705" alt="Screen Shot 2021-12-01 at 5 52 53 PM" src="https://user-images.githubusercontent.com/872351/144327967-6a417d88-e80f-4f87-8685-2179626b7734.png">

amalone-scwx commented:
LGTM

aleksmaus commented:
cla/check

aleksmaus commented:
@directionless @amalone-scwx is there a way to force CLA recheck? 
I was already added to the list of approved contributors by our CLA manager.

aleksmaus commented:
@zwass Thanks! Looks like still blocked on CLA verification. Will check with our team.

zwass commented:
Thought that might help! I don't seem to have admin access to our EasyCLA account so not sure if there's anything else I can do at the moment.

aleksmaus commented:
/easycla

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,203 @@
+
+#include <string>
+
+#include <atlbase.h>
```
Related: https://github.com/osquery/osquery/pull/7407#discussion_r767157102

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,203 @@
+
+HRESULT populateRow(INetFwRule* rule, Row& r);
+
+QueryData genFirewallRules(QueryContext& context) {
```
This function should make sure of the `Expected<Value, Error>` (`utils/expected/expected.h`) class and either return an error or the requested data in the form of a list of structures (without generating rows).

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,203 @@
+  return results;
+}
+
+HRESULT populateRow(INetFwRule* rule, Row& r) {
```
This function should receive a list of structures from `genFirewallRules` and render them to `QueryResults`. This makes it easy to add tests specifically targeting rendering

aleksmaus replied:
That would be less efficient, would take more memory and more copying data around, but sure can change.

alessandrogario replied:
> That would be less efficient, would take more memory and more copying data around, but sure can change.

This approach is preferred because it allows for better testing. Some tables return too much data for this approach, so we usually skip it there (example: tables accessing system logs)

aleksmaus replied:
I found it hard to find any table implemented with the suggested approach in the current osquery code, vast majority do not follow this pattern, so this makes it very inconsistent. 
Currently on my development pristine Windows 10 VM there are 470 firewall rules.
Anyways, will update this PR shortly as you requested, just to get this change in.

alessandrogario commented on tests/integration/tables/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,52 @@
+
+  ASSERT_EQ(data.size(), 1ul);
+
+  ValidationMap row_map = {
```
Once the code has been refactored, a unit test with mocked input data should make sure that the rendering works as expected with different input values

alessandrogario reviewed (changes requested):
Hello @aleksmaus,

thanks for the PR! I have left some comments

aleksmaus commented:
> Hello @aleksmaus,
> 
> thanks for the PR! I have left some comments

Thanks for the feedback! Most of the patterns borrowed from the existing code to make it consistent. Will update for suggested changes.

aleksmaus commented:
@alessandrogario I've made the changes that you requested. Could you take a look if this matches what you meant?
Also let me know if need to get rid of ATL.

directionless commented:
It looks like the CLA resolved itself. Closing and reopening is the simple way to get it to check.

FWIW a common issue is having commits that aren't tied to your github username in the PR. But for more complex stuff I usually dig up the easyCLA support email address.

aleksmaus commented on osquery/tables/networking/CMakeLists.txt:
```diff
@@ -116,6 +116,7 @@ function(generateOsqueryTablesNetworking)
   elseif(DEFINED PLATFORM_WINDOWS)
     list(APPEND public_header_files
       windows/interfaces.h
+      windows/windows_firewall_rules.h
```
@alessandrogario don't really like that I had to expose this header for requested testability.
let me know, what you think. can rollback this and the previous commit.

aleksmaus replied:
I just prefer not to expose the implementation details in the public headers for testability sake only.

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.h:
```diff
@@ -0,0 +1,38 @@
+namespace osquery {
+namespace tables {
+
+struct FirewallRule {
```
is_standard_layout/is_trivial types should be default-initialized

aleksmaus replied:
btw plenty of code where this is not done, example
https://github.com/osquery/osquery/blob/2e1fff970274d875cfe345b6534d29a0c6397d58/osquery/utils/windows/shelllnk.h

alessandrogario replied:
> btw plenty of code where this is not done, example https://github.com/osquery/osquery/blob/2e1fff970274d875cfe345b6534d29a0c6397d58/osquery/utils/windows/shelllnk.h

Yes, those structures should be fixed too (not in this PR)

alessandrogario commented on tests/integration/tables/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,242 @@
+namespace osquery {
+
+namespace tables {
+extern QueryData renderFirewallRules(const FirewallRules& rules);
```
This function should be exported from the public header

aleksmaus replied:
done

alessandrogario commented on tests/integration/tables/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,242 @@
+  ASSERT_EQ(rule.remotePorts, row["remote_ports"]);
+  ASSERT_EQ(rule.icmpTypesCodes, row["icmp_types_codes"]);
+
+  ASSERT_EQ(INTEGER(bool(rule.profileBitmask & NET_FW_PROFILE2_DOMAIN)),
```
C-style casts

aleksmaus replied:
updated

alessandrogario commented on tests/integration/tables/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,242 @@
+
+} // namespace
+
+TEST_F(windows_firewall_rules, test_firewall_rules_render) {
```
Unit tests are located near the implementation; in this case, the path is the following: `osquery/tables/networking/tests/CMakeLists.txt`

aleksmaus replied:
moved

alessandrogario reviewed (changes requested):
This is really good code @aleksmaus! I have left some additional comments, I think we can merge this soon

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,281 @@
+  if (FAILED(
+          hr = getString(rule, &INetFwRule::get_ApplicationName, &r.appName))) {
+    return createRuleCOMError(hr,
+                              "Failed to get firewall rule application name");
```
But you will not have the context where this COM error is coming from.

aleksmaus replied:
Has conversation over Slack over the exact details of the change, will update.

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,281 @@
+    break;
+  }
+
+  r["enabled"] = INTEGER(bool(rule.enabled));
```
borrowed this C cast from here
https://github.com/osquery/osquery/blob/d2be385d71f401c85872f00d479df8f499164c5a/osquery/tables/networking/windows/connectivity.cpp#L48

aleksmaus replied:
updated

alessandrogario replied:
> borrowed this C cast from here
> 
> https://github.com/osquery/osquery/blob/d2be385d71f401c85872f00d479df8f499164c5a/osquery/tables/networking/windows/connectivity.cpp#L48

The cast linked should be fixed too (but not in this PR)

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.h:
```diff
@@ -0,0 +1,38 @@
+  std::string localPorts;
+  std::string remotePorts;
+  std::string icmpTypesCodes;
+  long profileBitmask;
```
the API signature is long*
```
        virtual /* [propget][id] */ HRESULT STDMETHODCALLTYPE get_Profiles( 
            /* [retval][out] */ long *profileTypesBitmask) = 0;
```
https://docs.microsoft.com/en-us/windows/win32/api/netfw/nf-netfw-inetfwrule-get_profiles

aleksmaus replied:
the enum doesn't cover 0 value. 
```
typedef 
enum NET_FW_PROFILE_TYPE2_
    {
        NET_FW_PROFILE2_DOMAIN	= 0x1,
        NET_FW_PROFILE2_PRIVATE	= 0x2,
        NET_FW_PROFILE2_PUBLIC	= 0x4,
        NET_FW_PROFILE2_ALL	= 0x7fffffff
    } 	NET_FW_PROFILE_TYPE2;
```
which could be possible return value from the API

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.h:
```diff
@@ -0,0 +1,38 @@
+  NET_FW_ACTION action;
+  bool enabled;
+  NET_FW_RULE_DIRECTION direction;
+  long protocol;
```
The API signature is long*
```
HRESULT get_Protocol(
  LONG *protocol
);
```
https://docs.microsoft.com/en-us/windows/win32/api/netfw/nf-netfw-inetfwrule-get_protocol

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,281 @@
+        hr, "Failed to get firewall rules enumerator interface");
+  }
+
+  do {
```
done

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,281 @@
+
+    if (hr != S_FALSE) {
+      hr = value.ChangeType(VT_DISPATCH);
+      if (FAILED(hr)) {
```
done

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,281 @@
+
+      CComPtr<INetFwRule> rule;
+      hr = (V_DISPATCH(&value))->QueryInterface(IID_PPV_ARGS(&rule));
+      if (FAILED(hr)) {
```
done

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,281 @@
+}
+
+typedef HRESULT (STDMETHODCALLTYPE INetFwRule::*BSTRFunc)(BSTR*);
+HRESULT getString(INetFwRule* rule, BSTRFunc fn, std::string* s) {
```
In my previous lives on some projects people preferred pointers cause it's visible that the variable is changed by the function at the time of the call when passing param as &str vs str.
Adjusted.

aleksmaus commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,281 @@
+  }
+
+  r["profile_domain"] =
+      INTEGER(bool(rule.profileBitmask & NET_FW_PROFILE2_DOMAIN));
```
updated

alessandrogario reviewed (changes requested):
We have concluded that ATL should be removed since it does require a valid Visual Studio license, and not everyone contributing to osquery is eligible to use the Community Edition

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,411 @@
+// Implementation
+namespace {
+
+// Map lookup for error messages per code review request
```
This can be removed

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,411 @@
+  if (FAILED(hr)) {
+    r = createWindowsFirewallRulesError(WindowsFirewallError::NetFwPolicyError,
+                                        hr);
+    goto Cleanup;
```
`goto` statements should not be used (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-goto).

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,411 @@
+
+Cleanup:
+
+  if (enumvar != NULL) {
```
Manual cleanup should be avoided (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e6-use-raii-to-prevent-leaks).

I think these types can be easily wrapped around `std::unique_ptr` with a custom deleter. An example can be found here: `https://github.com/osquery/osquery/blob/master/osquery/tables/system/windows/intel_me.cpp#L260`

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,411 @@
+    enumvar->Release();
+  }
+
+  if (enumerator != NULL) {
```
Replace `NULL` values with `nullptr` (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-nullptr)

alessandrogario commented on osquery/tables/networking/windows/windows_firewall_rules.cpp:
```diff
@@ -0,0 +1,411 @@
+  return r;
+}
+
+typedef HRESULT (STDMETHODCALLTYPE INetFwRule::*BSTRFunc)(BSTR*);
```
`using` should be used instead of `typedef`: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-using

alessandrogario reviewed (dismissed):
This is excellent work, thank you for this cool table @aleksmaus!

aleksmaus commented:
> This is excellent work, thank you for this cool table @aleksmaus!

Thank you for the code reviews @alessandrogario!

directionless commented on specs/windows/windows_firewall_rules.table:
```diff
@@ -0,0 +1,22 @@
+description("Provides the list of Windows firewall rules.")
+schema([
+    Column("name", TEXT, "Friendly name of the rule"),
+    Column("app_name", TEXT, "Friendly name of the application to which the rule applies"),
```
Is there a non-friendly, but authoritative, way to link this rule to an app? (I haven't looked at the underlying API yet)

directionless replied:
Digging through https://docs.microsoft.com/en-us/windows/win32/api/netfw/nn-netfw-inetfwrule

I think `get_ServiceName` is worth adding, since that's documented as an optional parameter that may take the place of app name.

`get_grouping` might also be interesting

aleksmaus replied:
basically exposed this API: (https://docs.microsoft.com/en-us/windows/win32/api/netfw/nn-netfw-inetfwrule)
and got the description from https://docs.microsoft.com/en-us/windows/win32/api/netfw/nf-netfw-inetfwrule-get_applicationname

aleksmaus replied:
will add

aleksmaus commented:
Added ```grouping``` and ```service_name``` columns per @directionless request.

<img width="1182" alt="Screen Shot 2021-12-21 at 10 56 32 AM" src="https://user-images.githubusercontent.com/872351/146960800-0f62520b-099b-4b07-b35a-8437f7990e95.png">

<img width="2422" alt="Screen Shot 2021-12-21 at 10 57 12 AM" src="https://user-images.githubusercontent.com/872351/146960832-f54506e2-d718-4ca4-841e-c95ddde14909.png">

Smjert commented:
Closing and reopening so that it aligns with the latest master.

james-elastic commented:
Gentle bump, Is there anything else needed to get the pr merged?
